### PR TITLE
NCL-5058 Fix for ConcurrentModificationException

### PR DESCRIPTION
**Pull request by rnc**

### CODE REVIEW

**Comment by geoand** [comment]
Tests failed :(

**Comment by rnc** [comment]
retest this please

**Comment by rnc** [comment]
@geoand **finally** !!! Made the config reloadable and tweaked various other things :-)

File: manipulation/src/main/java/org/jboss/gm/manipulation/actions/MavenPublicationRepositoryAction.java [hunk 1]
```diff
@@ -43,7 +44,9 @@
     @Override
     public void execute(Project project) {
         if (!project.getPluginManager().hasPlugin("maven")) {
-            logger.warn("Legacy 'maven' plugin not detected, skipping publishing repository creation.");
+            // This should never happen due to prior checks in ManipulationPlugin
+            throw new ManipulationUncheckedException(
+                    "Legacy 'maven' plugin not detected, skipping publishing repository creation.");
```
Comment (rnc) [inline manipulation/src/main/java/org/jboss/gm/manipulation/actions/MavenPublicationRepositoryAction.java]: @geoand  I don't think this is even possible given the checks in ManipulationPlugin. Potentially the entire block could be removed.

File: manipulation/src/main/java/org/jboss/gm/manipulation/ManipulationPlugin.java [hunk 2]
```diff
@@ -50,6 +50,10 @@ public void apply(Project project) {
         final ManipulationModel alignmentModel = ManipulationIO.readManipulationModel(project.getRootDir());
         final ManipulationModel correspondingModule = alignmentModel.findCorrespondingChild(project);
 
+        if (System.getProperty("gmeFunctionalTest") != null) {
+            ConfigCache.getOrCreate(Configuration.class).reload();
+        }
+
         // we need to change the project version early so various tasks that ready early and create other vars based on it
```
Comment (rnc) [inline manipulation/src/main/java/org/jboss/gm/manipulation/ManipulationPlugin.java]: If we are running functional tests, trigger a reload.

File: analyzer/src/main/java/org/jboss/gm/analyzer/alignment/AlignmentPlugin.java [hunk 3]
```diff
@@ -42,12 +40,7 @@ public void apply(Project project) {
         }
 
         if (System.getProperty("gmeFunctionalTest") != null) {
-            final Configuration config = ConfigCache.getOrCreate(Configuration.class);
-            logger.info("Running functional test and injecting {}",
-                    System.getProperties().stringPropertyNames().stream()
-                            .filter(k -> config.getProperties().getProperty(k) == null).collect(Collectors.toList()));
-
-            config.getProperties().putAll(System.getProperties());
+            ConfigCache.getOrCreate(Configuration.class).reload();
```
Comment (geoand) [inline analyzer/src/main/java/org/jboss/gm/analyzer/alignment/AlignmentPlugin.java]: :100:

File: manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java [hunk 4]
```diff
@@ -54,10 +58,10 @@ public void ensureProperPomGenerated() throws IOException, URISyntaxException, X
 
         final ManipulationModel alignment = ManipulationIO.readManipulationModel(simpleProjectRoot);
 
-        final BuildResult buildResult = GradleRunner.create()
+        final BuildResult buildResult = TestUtils.createGradleRunner()
                 .withProjectDir(simpleProjectRoot)
-                .withArguments("generatePomFileForMainPublication")
-                .withDebug(true)
+                .withArguments("--info", "generatePomFileForMainPublication")
```
Comment (geoand) [inline manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java]: Do we want `--info`?

File: manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java [hunk 5]
```diff
@@ -96,24 +100,24 @@ public void ensureDocs() throws IOException, URISyntaxException {
...
+                .withArguments("--info", "publish")
                 .forwardOutput()
                 .withPluginClasspath()
+                //.withDebug(true)
                 .build();
 
         assertThat(buildResult.task(":" + "publishMainPublicationToJboss-snapshots-repositoryRepository").getOutcome())
                 .isEqualTo(TaskOutcome.SKIPPED);
-        assertThat(buildResult.task(":" + "publish").getOutcome()).isEqualTo(TaskOutcome.UP_TO_DATE);
+        assertThat(buildResult.task(":" + "publish").getOutcome()).isEqualTo(TaskOutcome.SUCCESS);
```
Comment (geoand) [inline manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java]: At last!

File: manipulation/src/main/java/org/jboss/gm/manipulation/ManipulationPlugin.java (hunk 2 above)
Comment (geoand) [reply manipulation/src/main/java/org/jboss/gm/manipulation/ManipulationPlugin.java]: :+1:

File: manipulation/src/main/java/org/jboss/gm/manipulation/actions/MavenPublicationRepositoryAction.java (hunk 1 above)
Comment (geoand) [reply manipulation/src/main/java/org/jboss/gm/manipulation/actions/MavenPublicationRepositoryAction.java]: Makes sense :ok_hand:

File: manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java (hunk 4 above)
Comment (rnc) [reply manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java]: *shrug* Fair point, I could remove it. Makes very little difference and was left in by mistake when debugging.

Comment (geoand) [reply manipulation/src/functTest/java/org/jboss/gm/manipulation/SimpleProjectFunctionalTest.java]: No problem 😊

**Pull request merged**